File: python/algo/202608/algo1117.py
import threading
from typing import Callable

class H2O:
    def __init__(self):
        self.lock = threading.Lock()
        self.sem = threading.Semaphore(2)
        self.bar = threading.Barrier(3)


    def hydrogen(self, releaseHydrogen: 'Callable[[], None]') -> None:
        with self.sem:
            # releaseHydrogen() outputs "H". Do not change or remove this line.
            releaseHydrogen()
            self.bar.wait()


    def oxygen(self, releaseOxygen: 'Callable[[], None]') -> None:
        with self.lock:
            # releaseOxygen() outputs "O". Do not change or remove this line.
            releaseOxygen()
            self.bar.wait()

File: cplus/leetcode/202608/algo1117.cpp
#include <bits/stdc++.h>
#include <barrier>
using namespace std;

class H2O {
private:
    // mutex mtx;
    counting_semaphore<2> sem_h{2};
    counting_semaphore<1> sem_o{1};
    barrier<> b{3};    
public:
    H2O() {
        
    }

    void hydrogen(function<void()> releaseHydrogen) {
        sem_h.acquire();
        // releaseHydrogen() outputs "H". Do not change or remove this line.
        releaseHydrogen();
        b.arrive_and_wait();
        sem_h.release();
    }

    void oxygen(function<void()> releaseOxygen) {
        // lock_guard lock{mtx};
        sem_o.acquire();
        // releaseOxygen() outputs "O". Do not change or remove this line.
        releaseOxygen();
        b.arrive_and_wait();
        sem_o.release();
    }
};